### DISCUSSION TITLE
ESPNow, water level sensor, deepsleep and extending system

### DISCUSSION THREAD

**Discussion by davefes** [micropython/micropython#10440] [ESP32] [answered]
I have a system https://github.com/davefes/ESP-Now-repeater and now would like to also put the repeater into light or deepsleep.

I would like suggestions on how to sync the system so that I can put not only the remote but also the repeater into light or deepsleep AND maintain downstream comms, ie every hour.  I can only think of sending the time upstream, say once a day and re-sync the repeater and remote to the local (monitoring) end.

The ESP32 system clock seems to be within a second or two a day, so I don't see hardware RTCs helping here.

Is there some technique out there that I am not aware of?

Thanks,
Dave

**Comment by karfas** [comment, marked as answer]
According to the ESP-IDF docs, you can also use use a 32kHz crystal as source of the RTC clock (see e.g. https://docs.espressif.com/projects/esp-idf/en/v4.4.3/esp32/api-reference/system/system_time.html?highlight=rtc%20clock)

Otherwise, I would sync the time every hour - when you already have communication - and (maybe) try to compensate for the known drift of the RTC clock. Also note that you don't need the correct time everywhere - you need only the same time on your repeater and on the nodes.

Using one of the MODEM_SLEEP modes during active time (have no idea if this works with ESPNow) might also help to maintain a longer communication window without too much power consumption.

**Reply by davefes** [reply]
Correction:  the 1 or 2 seconds I was referring to is when not using deep or lightsleep. I guess that the system clock uses the 8MHz xtal.

As I want to use deepsleep your suggestion of an external 32kHz seems easiest.  I have seen several circuits around with capacitors too ground. I don't see how to select 32K_XP and 32K_XN in the Quick reference for the ESP32.  Do I have to build my own firmware?

As getting a pressure reading once an hour is probably often enough I would have the opportunity to send the time upstream.  Correct ... the local (monitoring end) is mains powered and will be running all the time.

I will search for MODEM_SLEEP.

Thank you for the pointers

**Reply by davefes** [reply]
https://docs.micropython.org/en/latest/library/esp.html
Seems MODEM_SLEEP is only for ESP8266
https://lastminuteengineers.com/esp32-sleep-modes-power-consumption/
talks about modem sleep for the ESP32 but is that for Arduino?

**Comment by karfas** [comment]
In my opinion, it's a **requirement** for serious development in micropython to compile your own firmware. 
It requires months to years until the patches/pull requests you would like to try out **now** get merged into the main micropython branch.

The WLAN powersave option is e.g. in #8993, an alternate take in #6774.

**Reply by davefes** [reply]
After 20 years of compiling firmware for the ATmega series the real joy with Micropython is the ability to change code on-the-fly.  I was not a makefile power-user and my attempts at building firmware for the ESP8266 were unsuccessful.

Reading through some of the PRs I have often been overwhelmed by the depth of complexity and the knowledge that must be behind them!

**Reply by davefes** [reply]
I have noticed that when you come out of deepsleep that for a second or so you get a high current (40mA) spike.  I can see that will have some affect on overall current consumption in battery-constrained apps.  

Also, many beginners seem to get bitten by inadequate USB supply capabilities when the board boots-up.

**Reply by julien123123** [reply]
Are you aware if the spike you're describing is a micropython thing, an esp32 thing or a generalized deepsleep thing?

**Reply by davefes** [reply]
Possibly some WiFi init thing on the ESP32.  I have changed to LoRa for these sensors.  Maybe, start a new thread with your particular concern and see what pops-up.

**Comment by kjm1102** [comment]
A bit off topic I know Dave, but curious what you're using for a water level sensor? I've tried a couple of Aliexpress $40 stainless steel submersible depth probes but they invariably fail due water ingress after 6-12 months submerged.

**Reply by davefes** [reply]
Drop this into Aliexpress search:
`M15-2 M15-3 M15-4 M15-5 3m Cable Controller Float Switch Liquid Fluid Water Level Float Switch Controller Contactor Sensor`

**Reply by kjm1102** [reply]
The usual suspects. I wonder if the twice the price ones last twice as long?

**Reply by davefes** [reply]
:)  I use external 10K pull-ups just in case "wetting" current helps.  I would probably only get one closure per year, to tell me if water tanks are half empty.
